input/listen.go

```diff
@@ -6,42 +6,61 @@ import (
 )
 
 func listen(addr string, handler Handler) error {
-	laddr, err := net.ResolveTCPAddr("tcp", addr)
+	// listeners are set up outside of accept* here so they can interrupt startup
+	listener, err := listenTcp(addr)
 	if err != nil {
 		return err
 	}
-	l, err := net.ListenTCP("tcp", laddr)
+
+	udp_conn, err := listenUdp(addr)
 	if err != nil {
 		return err
 	}
-	log.Notice("listening on %v/tcp", laddr)
-	go acceptTcp(l, handler)
 
-	udp_addr, err := net.ResolveUDPAddr("udp", addr)
+	go acceptTcp(addr, listener, handler)
+	go acceptUdp(addr, udp_conn, handler)
+
+	return nil
+}
+
+func listenTcp(addr string) (*net.TCPListener, error) {
+	laddr, err := net.ResolveTCPAddr("tcp", addr)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	udp_conn, err := net.ListenUDP("udp", udp_addr)
+	listener, err := net.ListenTCP("tcp", laddr)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	log.Notice("listening on %v/udp", udp_addr)
-	go acceptUdp(udp_conn, handler)
-
-	return nil
+	return listener, nil
```

Comment:
instead of 
```
        listener, err := net.ListenTCP("tcp", laddr)
        if err != nil {
                return nil, err
        }
        return listener, nil
```
you can simply
```
return net.ListenTCP("tcp", laddr)
```

same for the UDP variant